Add test input component limited by maximum length.

aaronfc opened the pull request:
Fixes #5127

### Changes proposed in this Pull Request
* Add new `LimitedTextControl` component.
  * `maxLength` property defines the maximum length allowed.
  * `multiline` property controls if single line or multiline (textarea) input is used.
* Include sample usage in modal body (for testing purposes).

### Testing instructions
* Create a Lesson or Course.
* The modal should contain two inputs: one for title and one for description.
* Test that you can not go over the maximum length allowed.

### Screenshot / Video
<img width="948" alt="image" src="https://user-images.githubusercontent.com/799065/169052676-1256b492-4c2f-4a53-9a8d-2cd1f007bd3a.png">

### Discussion
- Initially I created two components `LimitedTextControl` and `LimitedTextareaControl` but being the only difference the base component I decided to refactor that to a `multiline` property.
- Label style currently does not fit the expected style. We can later implement a different label if we want.
- What's the difference between `useState` from `@wordpress/element` and `react` ?
- Open to suggestions on naming for the component.
- I am using Gutenberg's components (`TextControl` and `TextareaControl`) because I think the more "reusing" we do the better... but let me know if it might be better to just create com components from zero.

aaronfc committed:
Add `LimitedTextControl` component.

- Created `LimitedTextControl` based on `TextControl` and
  `TextareaControl` wordpress components.
- Numeric property `maxLength` controls the maximum length allowed for the
  input.
- Boolean property `multiline` controls if input or textarea shall be
  used when rendering.

aaronfc committed:
Update jsdoc for `LimitedTextControl` component.

renatho commented on assets/admin/editor-wizard/limited-text-control.js:
```diff
@@ -0,0 +1,47 @@
+	maxLength,
+	multiline = false,
+} ) => {
+	const changeValueIfAllowed = ( newValue ) => {
```
Quickly checking the `TextareaControl` and `TextControl`, I think they send the extra props we send to them to the `input` and `textarea`. Maybe we could simplify this by adding a `maxlength` prop to the `<Control>`?

aaronfc replied:
Much better! 👏  Thanks!

aaronfc committed:
Simplify logic for max length.

- Replace custom implementation with propagation of `maxlength` property
  throught the `maxLength` react property.

renatho reviewed (dismissed):
Looks good and works well!

Just one other thing that I thought is that maybe we could add this component as a generic one in `assets/blocks/editor-components`. But we can also do that in the future when needed.

aaronfc committed:
Move `LimitedTextControl` to `editor-components/`.

aaronfc commented:
Moved here 6f531fe9a 😄

aaronfc committed:
Add `LimitedTextControl` test.

aaronfc commented:
Bonus: Added some tests 😄

renatho reviewed (dismissed):
🎉

fjorgemota reviewed (dismissed):
Looks good and works well!

Just tested the new unit test, and also verified if copy-pasting didn't allow to break the limit (it doesn't). :+1:

aaronfc committed:
Refactor tests for `LimitedTextControl` component.

- Duplicate cases for `multiline = { true }`.
- Use `queryByRole` as much as possible.
- Add test for `Enter` key being allowed only on multiline case.
- Add test to ensure `maxLength` is respected.

aaronfc committed:
Add missing `multiline = true` to test.

aaronfc commented:
Sorry folks. I have refactored tests a little (still learning how to do proper tests in JS :P) and added cases for multiline.

Apparently I need again an approval 🤦 . No more changes, I promise!

fjorgemota commented on assets/blocks/editor-components/limited-text-control/index.test.js:
```diff
@@ -0,0 +1,180 @@
+import LimitedTextControl from './index';
+import userEvent from '@testing-library/user-event';
+
+describe( '<LimitedTextControl"/>', () => {
```
Any specific reason for adding this `"` here?

Sounds a little weird to have the name of the set of the tests as `<LimitedTextControl"/>`.

```suggestion
describe( '<LimitedTextControl />', () => {
```

fjorgemota commented on assets/blocks/editor-components/limited-text-control/index.test.js:
```diff
@@ -55,4 +54,127 @@ describe( '<LimitedTextControl />', () => {
+	} );
+} );
+
+describe( '<LimitedTextControl multiline="{ true }"/>', () => {
```
In the same sense, do we need the quotes here? 

Testing here, this works:

```suggestion
describe( '<LimitedTextControl multiline={ true }/>', () => {
```

fjorgemota reviewed (dismissed):
Nice new tests! 

Just left two comments about the quotes in the name of the tests, but nothing too important. =)

Thanks!

aaronfc committed:
Fix typo in test name.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

aaronfc committed:
Fix typo in test name.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -6,6 +6,11 @@ import { Modal } from '@wordpress/components';
+/**
+ * Internal dependencies
+ */
+import { __ } from '@wordpress/i18n';
```
Oops! I think it should be in the "WordPress dependencies", right? And we could take the opportunity to add a break line before the "Internal dependencies".

renatho commented on assets/blocks/editor-components/limited-text-control/index.test.js:
```diff
@@ -0,0 +1,180 @@
+
+		await userEvent.type( queryByRole( 'textbox' ), 'DEF' );
+
+		expect( onChangeMock ).toHaveBeenCalledTimes( 0 );
```
Thinking a little about this, I'd use a maxLength as `4`, and `toHaveBeenCalledTimes( 1 );`. So I think it's a little more tested. Otherwise, if `onChangeMock` is never called because it's not working, the test would pass. Does it make sense?

If you agree, I think the same applies to some of the other tests.

aaronfc replied:
I see your point, but I think we cover that case (when onChangeMock is never called) in the test "Should call the change event".

renatho commented on assets/blocks/editor-components/limited-text-control/index.test.js:
```diff
@@ -0,0 +1,180 @@
+
+		const element = queryByRole( 'textbox' );
+
+		await userEvent.type( element, '{enter}' );
```
Do we need the `async` / `await` here for some reason? The same question applies to the other tests.

aaronfc replied:
No. Apparently I thought it was based on some tests I did, but looks like I did my tests wrong 😓 
Removed here 876441ed0

aaronfc committed:
Update with `feature/course-lesson-wizard`.

aaronfc committed:
Remove unnecessary async/await in tests.

aaronfc commented:
After updating with `feature/course-lesson-wizard` I have moved the sample usage to the `CourseDetailsStep` step.
This is how it looks like now:
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/799065/169554614-6a45f39b-03da-4c81-a652-ea2c61a81d8a.png">

aaronfc committed:
Rename test case.

renatho reviewed (approved):
🚀
